xbmc/addons/AddonRepos.cpp

```diff
@@ -0,0 +1,220 @@
+  m_latestVersionsByRepo.clear();
+
+  const AddonRepoUpdateMode updateMode =
+      CAddonSystemSettings::GetInstance().GetAddonRepoUpdateMode();
+
+  for (const auto& repo : m_addonsByRepoMap)
+  {
+    const auto& addonsPerRepo = repo.second;
+
+    for (const auto& addonMapEntry : addonsPerRepo)
+    {
+      const auto& addonToAdd = addonMapEntry.second;
+
+      if (updateMode == AddonRepoUpdateMode::OFFICIAL_ONLY)
+      {
+        if (IsFromOfficialRepo(addonToAdd, true))
+        {
+          AddAddonIfLatest(addonToAdd, m_latestOfficialVersions);
+        }
+      }
+      else if (updateMode == AddonRepoUpdateMode::ANY_REPOSITORY)
+      {
+        AddAddonIfLatest(addonToAdd, m_latestOfficialVersions);
+      }
+
+      // add to latestVersionsByRepo - map
+      AddAddonIfLatest(repo.first, addonToAdd, m_latestVersionsByRepo);
+    }
+  }
+}
+
+void CAddonRepos::AddAddonIfLatest(const AddonPtr& addonToAdd,
+                                   std::map<std::string, AddonPtr>& map) const
+{
+  const auto& latestKnown = map.find(addonToAdd->ID());
+  if (latestKnown == map.end() || addonToAdd->Version() > latestKnown->second->Version())
+    map[addonToAdd->ID()] = addonToAdd;
+}
+
+void CAddonRepos::AddAddonIfLatest(
+    const std::string& repoId,
+    const AddonPtr& addonToAdd,
+    std::map<std::string, std::map<std::string, AddonPtr>>& map) const
+{
+  const auto& latestVersionByRepo = map.find(repoId);
+
+  if (latestVersionByRepo == map.end()) // repo not found
+  {
+    map[repoId].insert({addonToAdd->ID(), addonToAdd});
+  }
+  else
+  {
+    const auto& latestVersionEntryByRepo = latestVersionByRepo->second;
+    const auto& latestKnown = latestVersionEntryByRepo.find(addonToAdd->ID());
+
+    if (latestKnown == latestVersionEntryByRepo.end() ||
+        addonToAdd->Version() > latestKnown->second->Version())
+      map[repoId].insert({addonToAdd->ID(), addonToAdd});
+  }
+}
+
+void CAddonRepos::BuildUpdateList(const VECADDONS& installed, VECADDONS& updates) const
+{
+  for (const auto& addon : installed)
+  {
+    CLog::Log(LOGDEBUG, "ADDONS: update check: addonID = {} / Origin = {}", addon->ID(),
+              addon->Origin());
+
+    const AddonRepoUpdateMode updateMode =
+        CAddonSystemSettings::GetInstance().GetAddonRepoUpdateMode();
+
+    if (updateMode == AddonRepoUpdateMode::OFFICIAL_ONLY)
+    {
+      if (ORIGIN_SYSTEM == addon->Origin())
+      {
+        FindAddonAndCheckForUpdate(addon, GetLatestOfficialVersions(), updates);
+      }
+      else
+      {
+        const auto& repoEntry = GetLatestVersionsByRepo().find(addon->Origin());
+        if (repoEntry != GetLatestVersionsByRepo().end())
+        {
+          // own repository found, check for update
+          FindAddonAndCheckForUpdate(addon, repoEntry->second, updates);
+        }
+        else
+        {
+          // not found, looking at officials
+          FindAddonAndCheckForUpdate(addon, GetLatestOfficialVersions(), updates);
+        }
```

Comment:
We should always check the official repo first, correct? That use case should not change.

Reply:
So if there is no update in the official versions then then check the it's own repo.

Reply:
coming in with the next push